### PR TITLE
Refactor `Expr::Like`, `Expr::ILike`, `Expr::SimilarTo` to use a struct

### PR DESCRIPTION
**Pull request by b41sh**
# Which issue does this PR close?

Part of #2175

 # Rationale for this change

Refactor `Expr::Like`, `Expr::ILike`, `Expr::SimilarTo` to use a struct

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/expr/src/expr.rs [hunk 1]
```diff
@@ -101,26 +101,11 @@ pub enum Expr {
         right: Box<Expr>,
     },
     /// LIKE expression
-    Like {
-        negated: bool,
-        expr: Box<Expr>,
-        pattern: Box<Expr>,
-        escape_char: Option<char>,
-    },
+    Like(Like),
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:104]: That is a nice reduction in replication 👍

File: datafusion/expr/src/expr.rs [hunk 2]
```diff
@@ -580,39 +591,24 @@ impl Not for Expr {
...
-                expr,
-                pattern,
-                escape_char,
-            },
+            Expr::Like(like) => Expr::Like(Like::new(
+                !like.negated,
+                like.expr,
+                like.pattern,
+                like.escape_char,
+            )),
```
Comment (alamb) [inline datafusion/expr/src/expr.rs:599]: Another way you can write this type of match is like 

```suggestion
            Expr::Like(Like { negated, expr, pattern, escape_char} ) => Expr::Like(Like::new(
                negated,
                expr,
                pattern,
                escape_char,
            )),
```

The benefit of doing so is that if any new field is added to `Like` the compiler will force you to update all places that use `Like` to add the new field where as this way you would have to find them manually

**Review by alamb — APPROVED** [review]
LGTM -- thank you @b41sh 

I have a suggestion on style but I don't think it is required.

**Comment by andygrove** [comment]
Thanks @b41sh!

**Pull request merged**